test/range_test.exs

```diff
@@ -557,5 +655,107 @@ defmodule SourcerorTest.RangeTest do
                end: [line: 4, column: 7]
              }
     end
+
+    test "captures" do
+      assert to_range(~S"&foo/1") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 7]
+             }
+
+      assert to_range(~S"&Foo.bar/1") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 11]
+             }
+
+      assert to_range(~S"&__MODULE__.Foo.bar/1") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 22]
+             }
+    end
+
+    test "captures with arguments" do
+      assert to_range(~S"&foo(&1, :bar)") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 15]
+             }
+
+      assert to_range(~S"& &1.foo") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 9]
+             }
+
+      assert to_range(~S"& &1") == %{
+               start: [line: 1, column: 1],
+               end: [line: 1, column: 5]
+             }
+
+      # This range currently ends on column 5, though it should be column 6,
+      # and appears to be a limitation of the parser, which does not include
+      # any metadata about the parens. That is, this currently holds:
+      #
+      #     Sourceror.parse_string!("& &1") == Sourceror.parse_string!("&(&1)")
+      #
+      # assert to_range(~S"&(&1)") == %{
+      #          start: [line: 1, column: 1],
+      #          end: [line: 1, column: 6]
+      #        }
```

Comment:
Maybe we can propose a change to the elixir parser? Something similar used to happen for qualified module aliases like
```
Foo
.
   Bar
```
Which is a super contrived example but still valid, we added the `last` metadata to allow range calculation on them

Reply:
Good idea! I can post an issue.

Reply:
I think this may actually just be a bug. If you look at the underlying AST, `&` is just a call and should have a `:closing` metadata when parens are present:

```elixir
iex(1)> Code.string_to_quoted!("foo(bar)", token_metadata: true)
{:foo, [closing: [line: 1], line: 1], [{:bar, [line: 1], nil}]}

iex(2)> Code.string_to_quoted!("foo bar", token_metadata: true)
{:foo, [line: 1], [{:bar, [line: 1], nil}]}

iex(3)> Code.string_to_quoted!("&(&1)", token_metadata: true)
{:&, [line: 1], [{:&, [line: 1], [1]}]}

iex(4)> Code.string_to_quoted!("& &1", token_metadata: true)
{:&, [line: 1], [{:&, [line: 1], [1]}]}
```

Reply:
https://github.com/elixir-lang/elixir/issues/12919